ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +175,415 @@ let order_f (_, vm_rec) =
+let compute_anti_aff_evac_plan ~__context total_hosts hosts vms =
   let config =
     {Binpack.hosts; vms; placement= []; total_hosts; num_failures= 1}
   in
   Binpack.check_configuration config ;
-  debug "Planning configuration for offline agile VMs = %s"
-    (Binpack.string_of_configuration
-       (fun x ->
-         Printf.sprintf "%s (%s)" (Ref.short_string_of x)
-           (Db.Host.get_hostname ~__context ~self:x)
-       )
-       (fun x ->
-         Printf.sprintf "%s (%s)" (Ref.short_string_of x)
-           (Db.VM.get_name_label ~__context ~self:x)
-       )
-       config
-    ) ;
+
+  let binpack_plan ~__context config vms =
+    debug "Binpack planning configuration = %s"
+      (Binpack.string_of_configuration
+         (fun x ->
+           Printf.sprintf "%s (%s)" (Ref.short_string_of x)
+             (Db.Host.get_name_label ~__context ~self:x)
+         )
+         (fun x ->
+           Printf.sprintf "%s (%s)" (Ref.short_string_of x)
+             (Db.VM.get_name_label ~__context ~self:x)
+         )
+         config
+      ) ;
+    debug "VMs to attempt to evacuate: [ %s ]"
+      (String.concat "; "
+         (vms
+         |> List.map (fun (self, _) ->
+                Printf.sprintf "%s (%s)" (Ref.short_string_of self)
+                  (Db.VM.get_name_label ~__context ~self)
+            )
+         )
+      ) ;
+    let h = Binpack.choose_heuristic config in
+    h.Binpack.get_specific_plan config (List.map fst vms)
+  in
+
+  let binpack_after_plan_applied plan not_planned_vms =
+    match plan with
+    | [] ->
+        None
+    | plan -> (
+        debug "Binpack for the rest VMs" ;
+        let config_after_plan_applied = Binpack.apply_plan config plan in
+        let config_after_plan_applied =
+          {config_after_plan_applied with vms= not_planned_vms}
+        in
+        let b_plan =
+          binpack_plan ~__context config_after_plan_applied not_planned_vms
+        in
+        match List.length b_plan = List.length not_planned_vms with
+        | true ->
+            debug "Got final plan." ;
+            Some (plan @ b_plan)
+        | false ->
+            debug
+              "Failed to get final plan as failed to binpack for all the rest \
+               VMs." ;
+            None
+      )
+  in
+
+  match total_hosts with
+  | h when h < 3 ->
+      debug
+        "There are less than 2 available hosts to migrate VMs to, \
+         anti-affinity evacuation plan is not needed." ;
+      binpack_plan ~__context config vms
+  | _ ->
+      let anti_aff_vms, non_anti_aff_vms = vms |> vms_partition ~__context in
+      let spread_evenly_plan_pool_state =
+        init_spread_evenly_plan_pool_state ~__context anti_aff_vms hosts
+      in
+      let anti_aff_vms_increasing =
+        anti_aff_vms |> List.sort (fun (_, a, _) (_, b, _) -> compare a b)
+      in
+
+      let ( let* ) o f = match o with None -> f None | p -> p in
+      (let* _no_plan =
```

Comment:
It's `failed to plan`, and has to continue with the next step.
Otherwise, the plan will be returned directly.
This is the opposite of usual pattern `succeeded and then continue`.